ckanext/validation/utils.py

```diff
@@ -1,73 +1,304 @@
+    elif schema_json:
+        data_dict[u'schema'] = schema_json
 
-    This assumes the default folder structure of:
+    return data_dict
 
-        {ckan.storage_path}/resources/{uuid[0:3]}/{uuid[3:6}/{uuid[6:]}
 
-    Note: some checks are performed in case a custom uploader class is used,
-    but is not guaranteed to work in all circumstances. Please test before!
-    '''
-    path = get_local_upload_path(resource_id)
+def _get_missing_fields_from_current(current, new):
+    new['package_id'] = current['package_id']
+
+
+def _validate_resource(context, data_dict, new_resource=False):
+    mode = get_create_mode() if new_resource else get_update_mode()
+
+    if mode == "sync":
+        run_sync_validation(data_dict)
+    elif mode == "async":
+        run_async_validation(data_dict["id"])
+    else:
+        log.error("Validation mod isn't set. Skipping validation")
+
+
+def run_sync_validation(resource_data):
+    """If we are using sync validation (validation on update/create resource)
+    We must do it before the actual file upload, because if file is invalid
+    we don't want to replace the old one
+
+    Args:
+        resource_data (dict): new/updated resource data
+    """
+    schema = resource_data.get('schema')
+
+    if asbool(resource_data.get('align_default_schema')):
+        schema = _get_default_schema(resource_data["package_id"])
+
+    if schema and isinstance(schema, string_types):
+        schema = json.loads(schema)
+
+    _format = resource_data.get('format', '').lower()
+    options = _get_resource_validation_options(resource_data)
+    new_file = resource_data.get('upload')
+
+    if new_file:
+        source = _get_new_file_stream(new_file)
+    else:
+        if resource_data.get('url_type') == u'upload':
+            upload = get_resource_uploader(resource_data)
+
+            if isinstance(upload, ResourceUpload):
+                source = upload.get_path(resource_data["id"])
+            else:
+                source = resource_data['url']
+        else:
+            source = resource_data['url']
+
+    report = validate(source,
+                      format=_format,
+                      schema=schema,
+                      http_session=_get_session(resource_data),
+                      **options)
+
+    if report and not report['valid']:
+        for table in report.get('tables', []):
+            table['source'] = resource_data['url']
+
+        raise tk.ValidationError({u'validation': [report]})
+    else:
+        _table_count = report.get('table-count', 0) > 0
+
+        resource_data['validation_status'] = StatusTypes.success if _table_count else ""
+        resource_data['validation_timestamp'] = str(dt.now()) if _table_count else ""
+        resource_data['_success_validation'] = True
+
+
+def _get_resource_validation_options(resource_data):
+    options = config.get(u'ckanext.validation.default_validation_options')
+
+    if options:
+        options = json.loads(options)
```

Comment:
It's ugly to reuse the same variable for a completely different data type.

Also, this could become one line, eg `options = json.loads(options_str or '{}')`

Reply:
\+

Reply:
This will be fixed and pushed by Alex shortly.